### PR TITLE
Feature/f2018071301 add option for warnings per pr review

### LINKED ISSUE #302: f2018071301 - Add option for warnings per pr review
I would like to suggest a feature that will allow **limiting** the number of comment per **PR** review.
Currently, we are experiencing issues creating **PR reviews** in case the comments to publish are more (ex **100** etc). We are using the `github_pr_review` to publish comments to a **Github Enterprise** PR.

I suspect the **GHE** is not able to serve this request and returns **502**.
I believe this option will allow us to split the **PR** reviews and have smaller **PR** reviews.

-------

<details>
<summary>Backtrace</summary>

```ruby
PRONTO_GITHUB_ACCESS_TOKEN=access_token_here PRONTO_PULL_REQUEST_ID=1901 bin/pronto run -f github_pr_review -c origin/develop

warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
/Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/octokit-4.8.0/lib/octokit/response/raise_error.rb:16:in `on_complete': POST https://github-url.com/api/v3/repos/slug-here/pulls/1901/reviews: 502 - <!DOCTYPE html> (Octokit::BadGateway)
...
from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/faraday-0.15.2/lib/faraday/response.rb:9:in `block in call'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/faraday-0.15.2/lib/faraday/response.rb:61:in `on_complete'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/faraday-0.15.2/lib/faraday/response.rb:8:in `call'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/octokit-4.8.0/lib/octokit/middleware/follow_redirects.rb:73:in `perform_with_redirection'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/octokit-4.8.0/lib/octokit/middleware/follow_redirects.rb:61:in `call'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/faraday-0.15.2/lib/faraday/rack_builder.rb:143:in `build_response'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/faraday-0.15.2/lib/faraday/connection.rb:387:in `run_request'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/faraday-0.15.2/lib/faraday/connection.rb:175:in `post'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/sawyer-0.8.1/lib/sawyer/agent.rb:94:in `call'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/octokit-4.8.0/lib/octokit/connection.rb:156:in `request'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/octokit-4.8.0/lib/octokit/connection.rb:28:in `post'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/octokit-4.8.0/lib/octokit/client/reviews.rb:93:in `create_pull_request_review'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto/github.rb:51:in `create_pull_request_review'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto/formatter/github_pull_request_review_formatter.rb:13:in `submit_comments'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto/formatter/git_formatter.rb:9:in `format'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto.rb:67:in `block in run'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto.rb:66:in `each'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto.rb:66:in `run'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto/cli.rb:61:in `block in run'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto/cli.rb:60:in `chdir'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/lib/pronto/cli.rb:60:in `run'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/thor-0.19.4/lib/thor/command.rb:27:in `run'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/thor-0.19.4/lib/thor/invocation.rb:126:in `invoke_command'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/thor-0.19.4/lib/thor.rb:369:in `dispatch'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/thor-0.19.4/lib/thor/base.rb:444:in `start'
	from /Users/ventsislav.dimitrov/.rvm/gems/ruby-x.x.x/gems/pronto-0.9.5/bin/pronto:6:in `<top (required)>'
	from bin/pronto:18:in `load'
	from bin/pronto:18:in `<main>'
```

</details>

### PR DESCRIPTION
**Pull request by dimitrovv**
* Add option to limit comments per **PR** review 
* Create separate **PR** reviews depending on the comments to post

Fixes #302

### CODE REVIEW

**Comment by dimitrovv** [comment]
@mmozuras @mknapik could someone review this PR, will appreciate it.

We need this, cause currently **pronto** is 💥 when trying to post comment to **GHE**

File: .travis.yml [hunk 1]
```diff
@@ -3,18 +3,12 @@ before_install: gem update --system
 dist: trusty
 language: ruby
 rvm:
-- 2.0.0
```
Comment (dimitrovv) [inline .travis.yml]: I have removed this, cause it tries to resolve **public_suffix** to **3.0.2**, which requires **Ruby version >= 2.1** 🔴 build [403456633](https://travis-ci.org/prontolabs/pronto/jobs/403456633)

File: .gitignore [hunk 2]
```diff
@@ -4,3 +4,4 @@ pkg/*
 .DS_Store
 Gemfile.lock
 coverage
+.idea
```
Comment (mmozuras) [inline .gitignore]: @dimitrovv could you remove this change from this PR? Let's keep all strictly about warnings per PR review :)

File: .travis.yml (hunk 1 above)
Comment (mmozuras) [reply .travis.yml]: @dimitrovv could you make a separate PR for Travis changes? And let's figure out that one first.

File: lib/pronto/config_file.rb [hunk 3]
```diff
@@ -31,6 +31,7 @@ class ConfigFile
       'runners' => [],
       'formatters' => [],
       'max_warnings' => nil,
+      'warnings_per_review' => nil,
```
Comment (mmozuras) [inline lib/pronto/config_file.rb]: @dimitrovv maybe we should have a high default? What do you think?

File: lib/pronto/github.rb [hunk 4]
```diff
@@ -45,17 +45,13 @@ def create_pull_comment(comment)
...
-      options = {
-        event: 'COMMENT',
-        accept: 'application/vnd.github.black-cat-preview+json', # https://developer.github.com/v3/pulls/reviews/#create-a-pull-request-review
-        comments: comments.map do |c|
-          { path: c.path, position: c.position, body: c.body }
-        end
-      }
-      client.create_pull_request_review(slug, pull_id, options)
+    def publish_pull_request_comments(comments)
+      comments_left       = comments.clone
```
Comment (mmozuras) [inline lib/pronto/github.rb]: @dimitrovv we prefer a style of `comments_left_ = comments.clone` instead of aligning it this way.

File: lib/pronto/formatter/github_pull_request_review_formatter.rb [hunk 5]
```diff
@@ -10,7 +10,7 @@ def pretty_name
       end
 
       def submit_comments(client, comments)
-        client.create_pull_request_review(comments)
+        client.publish_pull_request_comments(comments)
```
Comment (mmozuras) [inline lib/pronto/formatter/github_pull_request_review_formatter.rb]: @dimitrovv why limit this feature to GitHub? Just trying to understand :)

File: lib/pronto/config_file.rb (hunk 3 above)
Comment (dimitrovv) [reply lib/pronto/config_file.rb]: yes, it is possible, I was 🤔not to force this for all so someone can continue using the current flow.
Ex, if not supplied, it will be nil and the comments will not be separated into different PRs. However we've noticed Github is taking too much time and no review is posted if many comments are submitted at once, so we could change it to 30 I guess - would be enough, what do you think?

File: lib/pronto/formatter/github_pull_request_review_formatter.rb (hunk 5 above)
Comment (dimitrovv) [reply lib/pronto/formatter/github_pull_request_review_formatter.rb]: yes, I thought about this, but I guess a review could only be submitted when using **Github** pronto client or I am wrong.

File: lib/pronto/github.rb (hunk 4 above)
Comment (dimitrovv) [reply lib/pronto/github.rb]: so I guess it would be better to leave it this way _(without alignment)_ or:

```ruby
comments_left = comments.clone
warnings_per_review = @config.warnings_per_review || comments.size
```

Comment (dimitrovv) [reply lib/pronto/github.rb]: Should I fix the alignment everywhere?

File: .travis.yml (hunk 1 above)
Comment (dimitrovv) [reply .travis.yml]: I saw you have already fixed this in [3b981f2](https://github.com/prontolabs/pronto/commit/3b981f2ecff5cba27ddb89ea7022d3c5fb1cfd99)

**Comment by dimitrovv** [comment]
@mmozuras what do you think of this? Is it good and whether could be reviewed again / merged soon?

File: lib/pronto/formatter/github_pull_request_review_formatter.rb (hunk 5 above)
Comment (mmozuras) [reply lib/pronto/formatter/github_pull_request_review_formatter.rb]: 👍

File: lib/pronto/config_file.rb [hunk 6]
```diff
@@ -1,6 +1,7 @@
 module Pronto
   class ConfigFile
     DEFAULT_MESSAGE_FORMAT = '%{msg}'.freeze
+    DEFAULT_WARNINGS_PER_REVIEW = 30
```
Comment (mmozuras) [inline lib/pronto/config_file.rb]: @dimitrovv why is the default value 30? What's the GH rate limit?

Comment (dimitrovv) [reply lib/pronto/config_file.rb]: @mmozuras we are using **Github Enterprise** and with disabled rate limit, but if I try to post about 40 comments per review **GHE** returns **502**. I don't this is smth with the rate limit here, have not tried with Public Github. I suspect the server is not able to serve this request somehow. 

Thus I can disable the default value here, so the comments will be separated into different reviews only if a default value is specified via config or env var. What do you think?

**Comment by dimitrovv** [comment]
@mmozuras what do you think, can we 🚢this soon, we need it when we have bigger PRs.

**Comment by doomspork** [comment]
@dimitrovv let's get this PR cleaned up so there aren't conflicts and I can give it a final once over and merge.

**Comment by dimitrovv** [comment]
@doomspork, thanks, already done, please review when possible, we need this

File: README.md [hunk 7]
```diff
@@ -116,6 +116,16 @@ If you want review to appear on pull request diff, instead of comments:
 $ PRONTO_GITHUB_ACCESS_TOKEN=token pronto run -f github_pr_review -c origin/master
 ```
 
+All the comments will now not be published into a single PR review, but separated into a number of PRs.
```
Comment (doomspork) [inline README.md]: @dimitrovv my only feedback is the wording here which is a little confusing to me. We don't mean to say it will create new PRs but will only leave _N_ number of feedback per Pronto run, right?

If I create a PR with 100 errors, Pronto will report the first 30.
If I fix them, and push a new commit to my PR, Pronto will report the next 30.

Is that understanding correct? If so, I think we could rephrase this as such:

> In order to avoid rate limiting by providers, Pronto publishes a limited number of comments per execution. By default, this is limited to the first 30 reported issues. You can change the number of reported issues either through the `.pronto.yml` file or the `PRONTO_WARNINGS_PER_REVIEW` environment variable.

Comment (dimitrovv) [reply README.md]: > @dimitrovv my only feedback is the wording here which is a little confusing to me. 

Yes, I didn't define it the right way, will fix

---

> If I create a PR with 100 errors, Pronto will report the first 30.
If I fix them, and push a new commit to my PR, Pronto will report the next 30.

Actually, it will post all N comments, but separated into X number of PR reviews (X = N / PRONTO_WARNINGS_PER_REVIEW). So all the comments should be published to Github only with a single pronto run

Comment (doomspork) [reply README.md]: Thank you @dimitrovv

Comment (dimitrovv) [reply README.md]: @doomspork, I've just tried to formulate the explanations around the new setting a bit better. Could you take a look when possible?

**Comment by doomspork** [comment]
@prontolabs/core any final thoughts before I merge this?

**Comment by chris-john-hopkins** [comment]
Is it possible to disable the comments but fail the check? If I set the max_warnings to 0, the check passes even if there are rubocop errors.
